# Oakland 311 calls

This notebook is part of Assignment #6.

In [1]:
import pandas as pd
import altair as alt
import requests

/Users/KaylaDong/.pyenv/versions/3.8.5/envs/lecture_06-3.8.5/lib/python3.8/site-packages/pandas/compat/__init__.py:124: UserWarning: Could not import the lzma module. Your installed Python is incomplete. Attempting to use lzma compression will result in a RuntimeError.
  warnings.warn(msg)


## Download data

Download [311 call data](https://data.oaklandca.gov/Infrastructure/Service-requests-received-by-the-Oakland-Call-Cent/quth-gb8e) from the City of Oakland's Open Data Portal. 

In [48]:
url = 'https://data.oaklandca.gov/api/views/quth-gb8e/rows.csv?accessType=DOWNLOAD'
r = requests.get(url, allow_redirects=False)

# write the content of the request into a file called `oakland_311.csv`
open('oakland_311.csv', 'wb').write(r.content)

179908119

## Import the csv into a pandas dataframe

Assign the dataframe into a variable called oakland_311_original

In [4]:
oakland_311_original = pd.read_csv('oakland_311.csv')

## View data

View the first 5 rows and the last 5 rows of the dataframe.

In [5]:
oakland_311_original.head()

REQUESTID            DATETIMEINIT       SOURCE  \
0     927708  07/24/2019 07:34:04 AM  SeeClickFix   
1     930512  08/01/2019 01:38:16 PM  SeeClickFix   
2     928898  07/28/2019 09:20:50 AM  SeeClickFix   
3     930760  08/02/2019 10:13:45 AM        Phone   
4     923520  07/11/2019 08:45:50 AM  SeeClickFix   

                            DESCRIPTION REQCATEGORY  \
0           Parking - Abandoned Vehicle      POLICE   
1  Illegal Dumping � mattress/boxspring     ILLDUMP   
2    Human Waste in Public-Right-of-Way       OTHER   
3          City Bldg - Structural Issue   BLDGMAINT   
4         Streets - Potholes/Depression    STREETSW   

                                  REQADDRESS    STATUS  \
0   (37.83332450468423, -122.26187389355277)    CLOSED   
1   (37.80947700189105, -122.29480799834606)    CLOSED   
2  (37.759826001279855, -122.16106799755731)  REFERRED   
3   (37.74458491831174, -122.16914534319535)    CLOSED   
4  (37.780819708391476, -122.22030767672499)    CANCEL   

                             REFERREDTO          DATETIMECLOSED          SRX  \
0                                   NaN  09/30/2019 10:25:11 AM  6052877.252   
1                                   NaN  08/04/2019 10:38:05 AM  6043201.257   
2  ADVANCE CRIME SCENE, RESTORATION EST                     NaN  6081509.580   
3                                   NaN  10/28/2019 11:30:12 AM  6079075.972   
4                                   NaN                     NaN  6064527.852   

           SRY COUNCILDISTRICT BEAT      PROBADDRESS     City State  
0  2130683.728            CCD1  12X      479 45TH ST  Oakland    CA  
1  2122182.593            CCD3  02Y  1117 CHESTER ST  Oakland    CA  
2  2103393.102            CCD7  35X   8601 SENECA ST  Oakland    CA  
3  2097885.948            CCD7  32X  HOLLY MINI PARK  Oakland    CA  
4  2111345.419            CCD5  23X        E 18TH ST  Oakland    CA

In [6]:
oakland_311_original.tail()

REQUESTID            DATETIMEINIT       SOURCE  \
852668     921970  07/07/2019 11:57:45 AM  SeeClickFix   
852669     930717  08/02/2019 08:58:15 AM  SeeClickFix   
852670     928953  07/28/2019 04:50:48 PM  SeeClickFix   
852671     920576  07/01/2019 04:27:51 PM        Phone   
852672     919254  06/27/2019 10:24:31 AM  SeeClickFix   

                                       DESCRIPTION REQCATEGORY  \
852668                       Parking - Enforcement     PARKING   
852669                 Parking - Abandoned Vehicle      POLICE   
852670                 Parking - Abandoned Vehicle      POLICE   
852671  Illegal Dumping - debris, appliances, etc.     ILLDUMP   
852672               Street Light - Outage/Damaged  ELECTRICAL   

                                      REQADDRESS           STATUS REFERREDTO  \
852668  (37.796293496626504, -122.2725785428372)  GONE ON ARRIVAL        NaN   
852669  (37.81960600132651, -122.27702919438389)           CLOSED        NaN   
852670   (37.75789339999198, -122.1766678975261)           CLOSED        NaN   
852671   (37.7697750419514, -122.18324682258614)           CLOSED        NaN   
852672   (37.84077618000984, -122.2688993400894)           CLOSED        NaN   

                DATETIMECLOSED          SRX          SRY COUNCILDISTRICT BEAT  \
852668                     NaN  6049530.979  2117260.646            CCD3  01X   
852669  09/03/2019 09:14:03 AM  6048406.421  2125772.131            CCD3  07X   
852670  12/26/2019 01:18:26 PM  6076988.064  2102769.930            CCD6  30X   
852671  07/01/2019 08:08:55 PM  6075164.293  2107129.719            CCD6  30X   
852672  07/05/2019 02:57:56 PM  6050899.986  2133434.859            CCD1  11X   

                    PROBADDRESS     City State  
852668               308 4TH ST  Oakland    CA  
852669           2825 MARKET ST  Oakland    CA  
852670         7924 plymouth st  Oakland    CA  
852671  66TH AVE & BANCROFT AVE  Oakland    CA  
852672              764 56TH ST  Oakland    CA

Try to call the dataframe on its own.

In [7]:
oakland_311_original

REQUESTID            DATETIMEINIT       SOURCE  \
0          927708  07/24/2019 07:34:04 AM  SeeClickFix   
1          930512  08/01/2019 01:38:16 PM  SeeClickFix   
2          928898  07/28/2019 09:20:50 AM  SeeClickFix   
3          930760  08/02/2019 10:13:45 AM        Phone   
4          923520  07/11/2019 08:45:50 AM  SeeClickFix   
...           ...                     ...          ...   
852668     921970  07/07/2019 11:57:45 AM  SeeClickFix   
852669     930717  08/02/2019 08:58:15 AM  SeeClickFix   
852670     928953  07/28/2019 04:50:48 PM  SeeClickFix   
852671     920576  07/01/2019 04:27:51 PM        Phone   
852672     919254  06/27/2019 10:24:31 AM  SeeClickFix   

                                       DESCRIPTION REQCATEGORY  \
0                      Parking - Abandoned Vehicle      POLICE   
1             Illegal Dumping � mattress/boxspring     ILLDUMP   
2               Human Waste in Public-Right-of-Way       OTHER   
3                     City Bldg - Structural Issue   BLDGMAINT   
4                    Streets - Potholes/Depression    STREETSW   
...                                            ...         ...   
852668                       Parking - Enforcement     PARKING   
852669                 Parking - Abandoned Vehicle      POLICE   
852670                 Parking - Abandoned Vehicle      POLICE   
852671  Illegal Dumping - debris, appliances, etc.     ILLDUMP   
852672               Street Light - Outage/Damaged  ELECTRICAL   

                                       REQADDRESS           STATUS  \
0        (37.83332450468423, -122.26187389355277)           CLOSED   
1        (37.80947700189105, -122.29480799834606)           CLOSED   
2       (37.759826001279855, -122.16106799755731)         REFERRED   
3        (37.74458491831174, -122.16914534319535)           CLOSED   
4       (37.780819708391476, -122.22030767672499)           CANCEL   
...                                           ...              ...   
852668   (37.796293496626504, -122.2725785428372)  GONE ON ARRIVAL   
852669   (37.81960600132651, -122.27702919438389)           CLOSED   
852670    (37.75789339999198, -122.1766678975261)           CLOSED   
852671    (37.7697750419514, -122.18324682258614)           CLOSED   
852672    (37.84077618000984, -122.2688993400894)           CLOSED   

                                  REFERREDTO          DATETIMECLOSED  \
0                                        NaN  09/30/2019 10:25:11 AM   
1                                        NaN  08/04/2019 10:38:05 AM   
2       ADVANCE CRIME SCENE, RESTORATION EST                     NaN   
3                                        NaN  10/28/2019 11:30:12 AM   
4                                        NaN                     NaN   
...                                      ...                     ...   
852668                                   NaN                     NaN   
852669                                   NaN  09/03/2019 09:14:03 AM   
852670                                   NaN  12/26/2019 01:18:26 PM   
852671                                   NaN  07/01/2019 08:08:55 PM   
852672                                   NaN  07/05/2019 02:57:56 PM   

                SRX          SRY COUNCILDISTRICT BEAT  \
0       6052877.252  2130683.728            CCD1  12X   
1       6043201.257  2122182.593            CCD3  02Y   
2       6081509.580  2103393.102            CCD7  35X   
3       6079075.972  2097885.948            CCD7  32X   
4       6064527.852  2111345.419            CCD5  23X   
...             ...          ...             ...  ...   
852668  6049530.979  2117260.646            CCD3  01X   
852669  6048406.421  2125772.131            CCD3  07X   
852670  6076988.064  2102769.930            CCD6  30X   
852671  6075164.293  2107129.719            CCD6  30X   
852672  6050899.986  2133434.859            CCD1  11X   

                    PROBADDRESS     City State  
0                   479 45TH ST  Oakland    CA  
1               1117 CHESTER 

See more info on the dataframe

In [8]:
oakland_311_original.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 852673 entries, 0 to 852672
Data columns (total 16 columns):
 #   Column           Non-Null Count   Dtype  
---  ------           --------------   -----  
 0   REQUESTID        852673 non-null  int64  
 1   DATETIMEINIT     852673 non-null  object 
 2   SOURCE           852673 non-null  object 
 3   DESCRIPTION      852673 non-null  object 
 4   REQCATEGORY      852592 non-null  object 
 5   REQADDRESS       836786 non-null  object 
 6   STATUS           852672 non-null  object 
 7   REFERREDTO       70172 non-null   object 
 8   DATETIMECLOSED   610875 non-null  object 
 9   SRX              836786 non-null  float64
 10  SRY              836786 non-null  float64
 11  COUNCILDISTRICT  809829 non-null  object 
 12  BEAT             812291 non-null  object 
 13  PROBADDRESS      852196 non-null  object 
 14  City             852673 non-null  object 
 15  State            852673 non-null  object 
dtypes: float64(2), int64(1), object(13)
me

## Check the types of data

In [9]:
oakland_311_original.head()

REQUESTID            DATETIMEINIT       SOURCE  \
0     927708  07/24/2019 07:34:04 AM  SeeClickFix   
1     930512  08/01/2019 01:38:16 PM  SeeClickFix   
2     928898  07/28/2019 09:20:50 AM  SeeClickFix   
3     930760  08/02/2019 10:13:45 AM        Phone   
4     923520  07/11/2019 08:45:50 AM  SeeClickFix   

                            DESCRIPTION REQCATEGORY  \
0           Parking - Abandoned Vehicle      POLICE   
1  Illegal Dumping � mattress/boxspring     ILLDUMP   
2    Human Waste in Public-Right-of-Way       OTHER   
3          City Bldg - Structural Issue   BLDGMAINT   
4         Streets - Potholes/Depression    STREETSW   

                                  REQADDRESS    STATUS  \
0   (37.83332450468423, -122.26187389355277)    CLOSED   
1   (37.80947700189105, -122.29480799834606)    CLOSED   
2  (37.759826001279855, -122.16106799755731)  REFERRED   
3   (37.74458491831174, -122.16914534319535)    CLOSED   
4  (37.780819708391476, -122.22030767672499)    CANCEL   

                             REFERREDTO          DATETIMECLOSED          SRX  \
0                                   NaN  09/30/2019 10:25:11 AM  6052877.252   
1                                   NaN  08/04/2019 10:38:05 AM  6043201.257   
2  ADVANCE CRIME SCENE, RESTORATION EST                     NaN  6081509.580   
3                                   NaN  10/28/2019 11:30:12 AM  6079075.972   
4                                   NaN                     NaN  6064527.852   

           SRY COUNCILDISTRICT BEAT      PROBADDRESS     City State  
0  2130683.728            CCD1  12X      479 45TH ST  Oakland    CA  
1  2122182.593            CCD3  02Y  1117 CHESTER ST  Oakland    CA  
2  2103393.102            CCD7  35X   8601 SENECA ST  Oakland    CA  
3  2097885.948            CCD7  32X  HOLLY MINI PARK  Oakland    CA  
4  2111345.419            CCD5  23X        E 18TH ST  Oakland    CA

Need to convert two columns: DATETIMEINIT and DATETIMECLOSED

In [10]:
oakland_311_original['DATETIMEINIT'].min()

'01/01/2011 05:51:45 PM'

In [ ]:
oakland_311_original['DATETIMEINIT'].max()

'12/31/2020 12:59:06 PM'

The operations might be wrong.

## Clean data part 1: Properly type the data

### Copy the original dataframe

In [12]:
oakland_311 = oakland_311_original.copy()

### Covert columns to datetime

In [13]:
oakland_311.head()

REQUESTID            DATETIMEINIT       SOURCE  \
0     927708  07/24/2019 07:34:04 AM  SeeClickFix   
1     930512  08/01/2019 01:38:16 PM  SeeClickFix   
2     928898  07/28/2019 09:20:50 AM  SeeClickFix   
3     930760  08/02/2019 10:13:45 AM        Phone   
4     923520  07/11/2019 08:45:50 AM  SeeClickFix   

                            DESCRIPTION REQCATEGORY  \
0           Parking - Abandoned Vehicle      POLICE   
1  Illegal Dumping � mattress/boxspring     ILLDUMP   
2    Human Waste in Public-Right-of-Way       OTHER   
3          City Bldg - Structural Issue   BLDGMAINT   
4         Streets - Potholes/Depression    STREETSW   

                                  REQADDRESS    STATUS  \
0   (37.83332450468423, -122.26187389355277)    CLOSED   
1   (37.80947700189105, -122.29480799834606)    CLOSED   
2  (37.759826001279855, -122.16106799755731)  REFERRED   
3   (37.74458491831174, -122.16914534319535)    CLOSED   
4  (37.780819708391476, -122.22030767672499)    CANCEL   

                             REFERREDTO          DATETIMECLOSED          SRX  \
0                                   NaN  09/30/2019 10:25:11 AM  6052877.252   
1                                   NaN  08/04/2019 10:38:05 AM  6043201.257   
2  ADVANCE CRIME SCENE, RESTORATION EST                     NaN  6081509.580   
3                                   NaN  10/28/2019 11:30:12 AM  6079075.972   
4                                   NaN                     NaN  6064527.852   

           SRY COUNCILDISTRICT BEAT      PROBADDRESS     City State  
0  2130683.728            CCD1  12X      479 45TH ST  Oakland    CA  
1  2122182.593            CCD3  02Y  1117 CHESTER ST  Oakland    CA  
2  2103393.102            CCD7  35X   8601 SENECA ST  Oakland    CA  
3  2097885.948            CCD7  32X  HOLLY MINI PARK  Oakland    CA  
4  2111345.419            CCD5  23X        E 18TH ST  Oakland    CA

In [14]:
oakland_311['DATETIMEINIT'] = pd.to_datetime(oakland_311['DATETIMEINIT'])

Compare DATETIMEINIT and DATETIMECLOSED

In [15]:
oakland_311.head()

REQUESTID        DATETIMEINIT       SOURCE  \
0     927708 2019-07-24 07:34:04  SeeClickFix   
1     930512 2019-08-01 13:38:16  SeeClickFix   
2     928898 2019-07-28 09:20:50  SeeClickFix   
3     930760 2019-08-02 10:13:45        Phone   
4     923520 2019-07-11 08:45:50  SeeClickFix   

                            DESCRIPTION REQCATEGORY  \
0           Parking - Abandoned Vehicle      POLICE   
1  Illegal Dumping � mattress/boxspring     ILLDUMP   
2    Human Waste in Public-Right-of-Way       OTHER   
3          City Bldg - Structural Issue   BLDGMAINT   
4         Streets - Potholes/Depression    STREETSW   

                                  REQADDRESS    STATUS  \
0   (37.83332450468423, -122.26187389355277)    CLOSED   
1   (37.80947700189105, -122.29480799834606)    CLOSED   
2  (37.759826001279855, -122.16106799755731)  REFERRED   
3   (37.74458491831174, -122.16914534319535)    CLOSED   
4  (37.780819708391476, -122.22030767672499)    CANCEL   

                             REFERREDTO          DATETIMECLOSED          SRX  \
0                                   NaN  09/30/2019 10:25:11 AM  6052877.252   
1                                   NaN  08/04/2019 10:38:05 AM  6043201.257   
2  ADVANCE CRIME SCENE, RESTORATION EST                     NaN  6081509.580   
3                                   NaN  10/28/2019 11:30:12 AM  6079075.972   
4                                   NaN                     NaN  6064527.852   

           SRY COUNCILDISTRICT BEAT      PROBADDRESS     City State  
0  2130683.728            CCD1  12X      479 45TH ST  Oakland    CA  
1  2122182.593            CCD3  02Y  1117 CHESTER ST  Oakland    CA  
2  2103393.102            CCD7  35X   8601 SENECA ST  Oakland    CA  
3  2097885.948            CCD7  32X  HOLLY MINI PARK  Oakland    CA  
4  2111345.419            CCD5  23X        E 18TH ST  Oakland    CA

In [16]:
oakland_311.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 852673 entries, 0 to 852672
Data columns (total 16 columns):
 #   Column           Non-Null Count   Dtype         
---  ------           --------------   -----         
 0   REQUESTID        852673 non-null  int64         
 1   DATETIMEINIT     852673 non-null  datetime64[ns]
 2   SOURCE           852673 non-null  object        
 3   DESCRIPTION      852673 non-null  object        
 4   REQCATEGORY      852592 non-null  object        
 5   REQADDRESS       836786 non-null  object        
 6   STATUS           852672 non-null  object        
 7   REFERREDTO       70172 non-null   object        
 8   DATETIMECLOSED   610875 non-null  object        
 9   SRX              836786 non-null  float64       
 10  SRY              836786 non-null  float64       
 11  COUNCILDISTRICT  809829 non-null  object        
 12  BEAT             812291 non-null  object        
 13  PROBADDRESS      852196 non-null  object        
 14  City             852

In [18]:
oakland_311['DATETIMECLOSED'] = pd.to_datetime(oakland_311['DATETIMECLOSED'])

In [19]:
oakland_311.head()

REQUESTID        DATETIMEINIT       SOURCE  \
0     927708 2019-07-24 07:34:04  SeeClickFix   
1     930512 2019-08-01 13:38:16  SeeClickFix   
2     928898 2019-07-28 09:20:50  SeeClickFix   
3     930760 2019-08-02 10:13:45        Phone   
4     923520 2019-07-11 08:45:50  SeeClickFix   

                            DESCRIPTION REQCATEGORY  \
0           Parking - Abandoned Vehicle      POLICE   
1  Illegal Dumping � mattress/boxspring     ILLDUMP   
2    Human Waste in Public-Right-of-Way       OTHER   
3          City Bldg - Structural Issue   BLDGMAINT   
4         Streets - Potholes/Depression    STREETSW   

                                  REQADDRESS    STATUS  \
0   (37.83332450468423, -122.26187389355277)    CLOSED   
1   (37.80947700189105, -122.29480799834606)    CLOSED   
2  (37.759826001279855, -122.16106799755731)  REFERRED   
3   (37.74458491831174, -122.16914534319535)    CLOSED   
4  (37.780819708391476, -122.22030767672499)    CANCEL   

                             REFERREDTO      DATETIMECLOSED          SRX  \
0                                   NaN 2019-09-30 10:25:11  6052877.252   
1                                   NaN 2019-08-04 10:38:05  6043201.257   
2  ADVANCE CRIME SCENE, RESTORATION EST                 NaT  6081509.580   
3                                   NaN 2019-10-28 11:30:12  6079075.972   
4                                   NaN                 NaT  6064527.852   

           SRY COUNCILDISTRICT BEAT      PROBADDRESS     City State  
0  2130683.728            CCD1  12X      479 45TH ST  Oakland    CA  
1  2122182.593            CCD3  02Y  1117 CHESTER ST  Oakland    CA  
2  2103393.102            CCD7  35X   8601 SENECA ST  Oakland    CA  
3  2097885.948            CCD7  32X  HOLLY MINI PARK  Oakland    CA  
4  2111345.419            CCD5  23X        E 18TH ST  Oakland    CA

### Now get the earliest and latest date of both columns

In [20]:
oakland_311['DATETIMEINIT'].min()

Timestamp('2009-07-01 08:05:36')

In [21]:
oakland_311['DATETIMEINIT'].max()

Timestamp('2021-10-31 01:08:19')

It shows that the data of year 2009 and 2021 is not complete. May need to exclude these two years later.

In [22]:
oakland_311['DATETIMECLOSED'].min()

Timestamp('2006-09-08 15:05:19')

In [23]:
oakland_311['DATETIMECLOSED'].max()

Timestamp('2021-10-29 16:24:38')

### Get the difference of the two date columns

Create a new column called Close_Time that shows how long it took for a case to be closed.

In [26]:
oakland_311['Close_Time'] = oakland_311['DATETIMECLOSED'] - oakland_311['DATETIMEINIT']

Check the result.

In [27]:
oakland_311.head()

REQUESTID        DATETIMEINIT       SOURCE  \
0     927708 2019-07-24 07:34:04  SeeClickFix   
1     930512 2019-08-01 13:38:16  SeeClickFix   
2     928898 2019-07-28 09:20:50  SeeClickFix   
3     930760 2019-08-02 10:13:45        Phone   
4     923520 2019-07-11 08:45:50  SeeClickFix   

                            DESCRIPTION REQCATEGORY  \
0           Parking - Abandoned Vehicle      POLICE   
1  Illegal Dumping � mattress/boxspring     ILLDUMP   
2    Human Waste in Public-Right-of-Way       OTHER   
3          City Bldg - Structural Issue   BLDGMAINT   
4         Streets - Potholes/Depression    STREETSW   

                                  REQADDRESS    STATUS  \
0   (37.83332450468423, -122.26187389355277)    CLOSED   
1   (37.80947700189105, -122.29480799834606)    CLOSED   
2  (37.759826001279855, -122.16106799755731)  REFERRED   
3   (37.74458491831174, -122.16914534319535)    CLOSED   
4  (37.780819708391476, -122.22030767672499)    CANCEL   

                             REFERREDTO      DATETIMECLOSED          SRX  \
0                                   NaN 2019-09-30 10:25:11  6052877.252   
1                                   NaN 2019-08-04 10:38:05  6043201.257   
2  ADVANCE CRIME SCENE, RESTORATION EST                 NaT  6081509.580   
3                                   NaN 2019-10-28 11:30:12  6079075.972   
4                                   NaN                 NaT  6064527.852   

           SRY COUNCILDISTRICT BEAT      PROBADDRESS     City State  \
0  2130683.728            CCD1  12X      479 45TH ST  Oakland    CA   
1  2122182.593            CCD3  02Y  1117 CHESTER ST  Oakland    CA   
2  2103393.102            CCD7  35X   8601 SENECA ST  Oakland    CA   
3  2097885.948            CCD7  32X  HOLLY MINI PARK  Oakland    CA   
4  2111345.419            CCD5  23X        E 18TH ST  Oakland    CA   

        Close_Time  
0 68 days 02:51:07  
1  2 days 20:59:49  
2              NaT  
3 87 days 01:16:27  
4              NaT

## Explore some operations on the new column

### Mean

In [30]:
oakland_311['Close_Time'].mean()

Timedelta('49 days 19:45:09.623400857')

### Median

In [31]:
oakland_311['Close_Time'].median()

Timedelta('4 days 22:03:37')

### Min

In [32]:
oakland_311['Close_Time'].min()

Timedelta('-2981 days +00:37:47')

-2981 days?? It doesn't make sense. Check if this is an error.

In [36]:
oakland_311[oakland_311['Close_Time'] < '0 days']

REQUESTID        DATETIMEINIT          SOURCE  \
123682     432312 2012-11-01 11:24:56  Phone or Email   
229329      23385 2010-09-16 13:59:19  Phone or Email   
229330      23396 2010-09-16 13:59:46  Phone or Email   
229429      24888 2011-04-06 16:33:06  Phone or Email   
229603      25105 2011-05-16 11:55:22  Phone or Email   
...           ...                 ...             ...   
441281     522752 2014-06-10 11:29:36  Phone or Email   
443441     522842 2014-06-10 15:50:45  Phone or Email   
444304     528272 2014-07-10 12:20:22     SeeClickFix   
447744     522805 2014-06-10 14:02:16  Phone or Email   
469286     549531 2014-11-06 14:27:32           Phone   

                                            DESCRIPTION REQCATEGORY  \
123682                                             PG&E       OTHER   
229329             Residential Recycling Service Issues   RECYCLING   
229330             Residential Recycling Service Issues   RECYCLING   
229429  Garbage - Requests Exemption - Weekly or Shared   RECYCLING   
229603             Residential Recycling Service Issues   RECYCLING   
...                                                 ...         ...   
441281             Recycling - Missed Pickup - Backyard   RECYCLING   
443441                  Recycling - Repair/Replace Cart   RECYCLING   
444304                              Hazardous Materials        FIRE   
447744                  Recycling - Repair/Replace Cart   RECYCLING   
469286             Illegal Dumping � mattress/boxspring     ILLDUMP   

                                       REQADDRESS  STATUS  \
123682  (37.790247170707254, -122.23014783691161)  CLOSED   
229329  (30.009899650831787, -141.21875098935422)  CLOSED   
229330  (30.009899650831787, -141.21875098935422)  CLOSED   
229429  (30.009899650831787, -141.21875098935422)  CLOSED   
229603  (30.009899650831787, -141.21875098935422)  CLOSED   
...                                           ...     ...   
441281   (37.81947378040759, -122.19864842876396)  CLOSED   
443441   (37.81389394379093, -122.25288667849271)  CLOSED   
444304  (37.823866701293426, -122.26193144921672)  CLOSED   
447744    (37.82652163613099, -122.2607234958911)  CLOSED   
469286  (37.823866701293426, -122.26193144921672)  CLOSED   

                                  REFERREDTO      DATETIMECLOSED         SRX  \
123682                                 PG&E, 2012-07-10 14:00:14  6061748.00   
229329                                   NaN 2010-09-16 13:59:07        0.00   
229330                                   NaN 2010-09-16 13:59:35        0.00   
229429                                   NaN 2011-04-06 16:33:04        0.00   
229603                                   NaN 2011-05-16 11:55:21        0.00   
...                                      ...                 ...         ...   
441281  CALIFORNIA WASTE SOLUTIONS, EXTERNAL 2014-06-06 16:12:25  6071041.76   
443441  CALIFORNIA WASTE SOLUTIONS, EXTERNAL 2014-06-06 16:12:25  6055339.74   
444304                 FIRE (GRIFFIN, LEROY) 2006-09-08 15:05:19  6052795.80   
447744  CALIFORNIA WASTE SOLUTIONS, EXTERNAL 2014-06-06 16:12:25  6053162.82   
469286                                   NaN 2006-09-08 15:05:19  6052795.80   

               SRY COUNCILDISTRICT BEAT          PROBADDRESS     City State  \
123682  2114830.00             NaN  20X  E 23RD ST & 23RD AV  Oakland    CA   
229329        0.00             NaN  NaN                   ZZ  Oakland    CA   
229330        0.00             NaN  NaN                   ZZ  Oakland    CA   
229429        0.00             NaN  NaN                   ZZ  Oakland    CA   
229603        0.00             NaN  NaN                   ZZ  Oakland    CA   
...            ...             ...  ...                  ...      ...   ...   
441281  2125303.60            CCD4  13Z    2300 MASTLANDS DR  Oakland    CA   
443441  2123561.09            CCD3  14X        302 EUCLID AV  Oakland    CA   
444304  2127240.83            CCD3  08X        MOSSWOOD PARK  Oak

All these all errors? Need to ask the city for more information.

### Max

In [37]:
oakland_311['Close_Time'].max()

Timedelta('4293 days 21:59:22')

This seems like way too long. There might still be an error. 

In [38]:
oakland_311[oakland_311['Close_Time'] >= '4293 days']

REQUESTID        DATETIMEINIT          SOURCE  \
264880     323595 2009-12-09 14:30:28  Phone or Email   

                        DESCRIPTION REQCATEGORY  \
264880  Street Medians (Landscaped)       PARKS   

                                      REQADDRESS  STATUS REFERREDTO  \
264880  (37.82206394742191, -122.20671960118034)  CLOSED        NaN   

            DATETIMECLOSED        SRX        SRY COUNCILDISTRICT BEAT  \
264880 2021-09-11 12:29:50  6068728.0  2126289.0            CCD4  22X   

                      PROBADDRESS     City State         Close_Time  
264880  PARK BLVD & MOUNTAIN BLVD  Oakland    CA 4293 days 21:59:22

May need to ask the city for more info as well.

### Sort data

Check the first 10 rows of sorted data.

In [39]:
oakland_311.sort_values(by=['Close_Time'], ascending=False).head(10)

REQUESTID        DATETIMEINIT          SOURCE  \
264880     323595 2009-12-09 14:30:28  Phone or Email   
273400     357227 2011-01-11 16:11:55         Website   
275671     356540 2011-01-05 21:40:17         Website   
302161     370528 2011-05-18 15:53:54  Phone or Email   
293324     370530 2011-05-18 15:59:39  Phone or Email   
287663     368757 2011-05-02 16:22:17  Phone or Email   
292899     368880 2011-05-03 13:01:49  Phone or Email   
297456     368848 2011-05-03 11:17:04  Phone or Email   
282843     357043 2011-01-10 14:44:07         Website   
267036     356595 2011-01-06 09:25:05         Website   

                                 DESCRIPTION   REQCATEGORY  \
264880           Street Medians (Landscaped)         PARKS   
273400                     Sidewalk - Damage  ROW_STREETSW   
275671  Streets/Sidewalks - Unsafe Condition  ROW_STREETSW   
302161          Park - Landscape Maintenance         PARKS   
293324          Park - Landscape Maintenance         PARKS   
287663  Streets/Sidewalks - Unsafe Condition  ROW_STREETSW   
292899  Streets/Sidewalks - Unsafe Condition  ROW_STREETSW   
297456                     Sidewalk - Damage  ROW_STREETSW   
282843                     Sidewalk - Damage  ROW_STREETSW   
267036                     Sidewalk - Damage  ROW_STREETSW   

                                       REQADDRESS  STATUS REFERREDTO  \
264880   (37.82206394742191, -122.20671960118034)  CLOSED        NaN   
273400   (37.73561875315405, -122.17677744021242)  CLOSED        NaN   
275671    (37.81062199910106, -122.2589020860214)  CLOSED        NaN   
302161   (37.75119515507376, -122.15561532358669)  CLOSED        NaN   
293324   (37.75119515507376, -122.15561532358669)  CLOSED        NaN   
287663  (37.813745661961015, -122.24378665464081)  CLOSED        NaN   
292899    (37.76896035736208, -122.1697033555222)  CLOSED        NaN   
297456  (37.810750150595624, -122.24425986126634)  CLOSED        NaN   
282843    (37.80045375593913, -122.2765999639812)  CLOSED        NaN   
267036   (37.812484999176895, -122.2534134035195)  CLOSED        NaN   

            DATETIMECLOSED          SRX          SRY COUNCILDISTRICT BEAT  \
264880 2021-09-11 12:29:50  6068728.000  2126289.000            CCD4  22X   
273400 2021-08-05 09:29:00  6076811.090  2094661.167            CCD7  31Z   
275671 2021-07-14 09:57:37  6053579.984  2122402.509            CCD3  14X   
302161 2021-09-11 12:08:19  6083030.000  2100223.000            CCD7  35X   
293324 2021-09-11 12:08:47  6083030.000  2100223.000            CCD7  35X   
287663 2021-08-23 15:18:27  6057967.000  2123458.000            CCD2  14Y   
292899 2021-08-03 13:02:52  6079073.000  2106763.000            CCD6  30Y   
297456 2021-07-26 13:07:31  6057810.000  2122370.000            CCD2  14Y   
282843 2021-03-25 13:02:36  6048398.000  2118797.250            CCD3  03Y   
267036 2021-03-18 13:21:37  6055178.000  2123051.000            CCD3  14X   

                      PROBADDRESS     City State         Close_Time  
264880  PARK BLVD & MOUNTAIN BLVD  Oakland    CA 4293 days 21:59:22  
273400            10306 PIPPIN ST  Oakland    CA 3858 days 17:17:05  
275671          325 PARK VIEW TER  Oakland    CA 3842 days 12:17:20  
302161            FIRE STATION 26  Oakland    CA 3768 days 20:14:25  
293324            FIRE STATION 26  Oakland    CA 3768 days 20:09:08  
287663            847 WARFIELD AV  Oakland    CA 3765 days 22:56:10  
292899                7521 NEY AV  Oakland    CA 3745 days 00:01:03  
297456          3333 LAKESHORE AV  Oakland    CA 3737 days 01:50:27  
282843      7TH STREET & CLAY ST.  Oakland    CA 3726 days 22:18:29  
267036                377 PALM AV  Oakland    CA 3724 days 03:56:32

## Clean data part 2: Check and vet the data

### Check if there's unique identifier for every row

In [40]:
oakland_311['REQUESTID'].nunique()

852673

In [41]:
len(oakland_311)

852673

There numbers of unique case IDs and rows in the dataframe are both 852673! There might not be duplicates.

### Subsetting

Still check for any existing duplicates.

In [42]:
oakland_311[oakland_311.duplicated()]

Empty DataFrame
Columns: [REQUESTID, DATETIMEINIT, SOURCE, DESCRIPTION, REQCATEGORY, REQADDRESS, STATUS, REFERREDTO, DATETIMECLOSED, SRX, SRY, COUNCILDISTRICT, BEAT, PROBADDRESS, City, State, Close_Time]
Index: []

No exact duplicates found.

In [43]:
oakland_311[oakland_311['REQUESTID'].duplicated()]

Empty DataFrame
Columns: [REQUESTID, DATETIMEINIT, SOURCE, DESCRIPTION, REQCATEGORY, REQADDRESS, STATUS, REFERREDTO, DATETIMECLOSED, SRX, SRY, COUNCILDISTRICT, BEAT, PROBADDRESS, City, State, Close_Time]
Index: []

Still no duplicates found.

### Assert

Double-check if the length of the dataframe now matches the number of unique IDs.

In [47]:
assert len(oakland_311) == oakland_311['REQUESTID'].nunique()

## Export a clean version of the data to a csv

create a new notebook and exclude the column of row ids

In [49]:
oakland_311.to_csv('oakland_311_clean.csv', index = False)